### PR TITLE
Cost statistics

### PR DESCRIPTION
**Pull request by RezenkovD**

### CODE REVIEW

File: crud/crud_months_statistics.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+from models.user import User
+from schemas.statistics import Statistics
+
+
+def get_user_statistics(db: Session, user_id: int, month: int):
+    if month <= 0:
+        raise HTTPException(status_code=404, detail="Enter a positive number")
+    if month > 12:
+        raise HTTPException(status_code=404, detail="There is no such month number")
+    user = db.query(User).filter(User.id == user_id).first()
```
Comment (qqlexa) [inline crud/crud_months_statistics.py:12]: Use `one_or_none` instead `first`

File: crud/crud_months_statistics.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from fastapi import HTTPException
+from sqlalchemy.orm import Session
+from models.user import User
+from schemas.statistics import Statistics
+
+
+def get_user_statistics(db: Session, user_id: int, month: int):
```
Comment (qqlexa) [inline crud/crud_months_statistics.py:7]: Month? User can save information only during a year? 
I think that filter by year-month should be here

File: crud/crud_months_statistics.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if month > 12:
+        raise HTTPException(status_code=404, detail="There is no such month number")
+    user = db.query(User).filter(User.id == user_id).first()
+    if user is None:
+        raise HTTPException(status_code=404, detail="User not found")
+    number_purchases = len(user.items)
+    count_items_month = 0
+    number_categories = len(user.categories)
+    list_categories = []
+    for x in range(number_categories):
```
Comment (qqlexa) [inline crud/crud_months_statistics.py:19]: `for category in user.catefories`?

File: crud/crud_months_statistics.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    details_dict = {category: 0 for category in list_categories}
+    number_purchases_category = {category: 0 for category in list_categories}
+    all_costs = 0
+    for x in range(number_purchases):
+        if int(user.items[x].time.strftime("%m")) == month:
+            details_dict[user.items[x].category.title] += user.items[x].price
+            all_costs += user.items[x].price
+            number_purchases_category[user.items[x].category.title] += 1
+            count_items_month += 1
+    user_stats = Statistics
```
Comment (qqlexa) [inline crud/crud_months_statistics.py:30]: Should it be class or class instance?
For example `Statistics()` is class instance (object)

File: crud/crud_statistics.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from fastapi import HTTPException
+from sqlalchemy.orm import Session
+from models.user import User
+from schemas.statistics import Statistics
+
+
+def get_user_statistics(db: Session, user_id: int):
+    user = db.query(User).filter(User.id == user_id).first()
```
Comment (qqlexa) [inline crud/crud_statistics.py:8]: `one_or_none`

Comment (qqlexa) [inline crud/crud_statistics.py:8]: Use filter_by when you have only one model in query

File: crud/crud_statistics.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+from fastapi import HTTPException
+from sqlalchemy.orm import Session
+from models.user import User
+from schemas.statistics import Statistics
+
+
+def get_user_statistics(db: Session, user_id: int):
```
Comment (qqlexa) [inline crud/crud_statistics.py:7]: Is it not duplicate?
What about DRY?

File: main.py [hunk 7]
```diff
@@ -56,6 +58,27 @@ def create_category_for_user(
...
-        db=db, category=category
+    return crud.crud_category.create_user_category(db=db, category=category)
+
+
+@app.get(
+    "/users/{user_id}/statistics", response_model=schemas.statistics.Statistics
+)
+def get_stats(user_id: int, db: Session = Depends(get_db)):
+    statistics = crud.crud_statistics.get_user_statistics(db, user_id=user_id)
+    if statistics is None:
```
Comment (qqlexa) [inline main.py:69]: Can get_user_statistic return `None` if User not found? No because of checking this on the module 10th line

File: main.py [hunk 8]
```diff
@@ -56,6 +58,27 @@ def create_category_for_user(
...
+        raise HTTPException(status_code=404, detail="User not found")
+    return statistics
+
+
+@app.get(
+    "/users/{user_id}/statistics/{month}",
+    response_model=schemas.statistics.Statistics,
+)
+def get_stats_month(user_id: int, month: int, db: Session = Depends(get_db)):
```
Comment (qqlexa) [inline main.py:78]: Why do not we use Pydantic validator for checking month?

File: main.py [hunk 9]
```diff
@@ -56,6 +58,27 @@ def create_category_for_user(
...
+    statistics_month = crud.crud_months_statistics.get_user_statistics(
+        db, user_id=user_id, month=month
     )
+    if statistics_month is None:
+        raise HTTPException(status_code=404, detail="User not found")
```
Comment (qqlexa) [inline main.py:83]: I am not sure this is useful because of duplicate checking in there function

File: schemas/statistics.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+class Statistics(BaseModel):
+    email: str
+    costs: float
+    number_purchases: int
+    details: dict
+    number_purchases_category: dict
+
+    class Config:
+        orm_mode = True
```
Comment (qqlexa) [inline schemas/statistics.py:12]: ??

**Review by qqlexa — CHANGES_REQUESTED** [review]

**Comment by qqlexa** [comment]
Branch name should be in kebab-case

File: crud/crud_months_statistics.py (hunk 2 above)
Comment (RezenkovD) [reply crud/crud_months_statistics.py:7]: Thank you. It is very important.

File: crud/crud_months_statistics.py (hunk 3 above)
Comment (RezenkovD) [reply crud/crud_months_statistics.py:19]: And really. Redundant code.

File: crud/crud_months_statistics.py (hunk 4 above)
Comment (RezenkovD) [reply crud/crud_months_statistics.py:30]: Object. My carelessness.

File: crud/crud_statistics.py (hunk 5 above)
Comment (RezenkovD) [reply crud/crud_statistics.py:8]: Thanks.

File: main.py (hunk 7 above)
Comment (RezenkovD) [reply main.py:69]: It's true.

File: main.py (hunk 9 above)
Comment (RezenkovD) [reply main.py:83]: I fixed it)

File: main.py (hunk 8 above)
Comment (RezenkovD) [reply main.py:78]: Good offer. Thanks.

**Pull request closed without merging**